### PR TITLE
CI: Added Terraform deployment in circle CI

### PR DESCRIPTION
**Pull request by njm3754**
Terraform deployment added into config.yml to deploy infrastructure

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -18,6 +18,25 @@ workflows:
           filters:
             branches:
               only: master
+  terraform:
```
Comment (tbarlow12) [inline .circleci/config.yml:21]: Can we be more specific on the job name? There will be other terraform-related jobs that run as quality gates as well

File: .circleci/config.yml [hunk 2]
```diff
@@ -60,3 +79,75 @@ jobs:
...
+    steps:
+      - attach_workspace:
+          at: .
+      - run:
+          name: terraform plan
+          command: |
+            cd terraform
+            terraform plan \
+              -var "location=${REGION}" \
+              -var "tags={environment="\"${ENV_DEV}"\"}" \
```
Comment (tbarlow12) [inline .circleci/config.yml:113]: What is `ENV_DEV`?

File: .circleci/config.yml [hunk 3]
```diff
@@ -18,6 +18,25 @@ workflows:
...
+      - validate:
+          requires: 
+            - plan
+      - hold-apply:
+          type: approval
+          requires:
+            - validate
+      - apply:
+          requires:
+            - hold-apply
```
Comment (tbarlow12) [inline .circleci/config.yml:39]: These seem more like steps instead of jobs... Is there a specific reason why you wanted jobs?

**Review by tbarlow12 — COMMENTED** [review]
It's a good start, just a few things I'd like to see addressed before we merge

File: .circleci/config.yml (hunk 1 above)
Comment (njm3754) [reply .circleci/config.yml]: Done

File: .circleci/config.yml (hunk 2 above)
Comment (njm3754) [reply .circleci/config.yml]: Done

File: .circleci/config.yml (hunk 3 above)
Comment (njm3754) [reply .circleci/config.yml]: Done

File: .circleci/config.yml [hunk 4]
```diff
@@ -18,6 +18,18 @@ workflows:
...
+      - terraform-init-plan-validate:
+          context: 
+            - tf-deploy-dev
+      - hold-terraform-apply:
+          type: approval
+          requires:
+            - terraform-init-plan-validate
+      - terraform-apply:
+          requires:
+            - hold-terraform-apply
```
Comment (tbarlow12) [inline .circleci/config.yml:32]: This will also likely need the context

File: .circleci/config.yml [hunk 5]
```diff
@@ -18,6 +18,18 @@ workflows:
           filters:
             branches:
               only: master
+  iac_deploy_terraform:
+    jobs:
+      - terraform-init-plan-validate:
+          context: 
+            - tf-deploy-dev
```
Comment (tbarlow12) [inline .circleci/config.yml:25]: This should only run on `main`

**Review by tbarlow12 — COMMENTED** [review]
Almost there

File: .circleci/config.yml [hunk 6]
```diff
@@ -21,7 +21,25 @@ workflows:
       - build_and_deploy:
           filters:
             branches:
-              only: master
+              only: main
+  iac_deploy_terraform:
+    jobs:
+      - terraform-init-plan-validate:
+          context: 
+            - tf-deploy-dev
```
Comment (wbreza) [inline .circleci/config.yml:29]: What do these `context` entries mean?

File: .circleci/config.yml [hunk 7]
```diff
@@ -80,3 +98,50 @@ jobs:
...
+    docker:
+      - image: docker.mirror.hashicorp.services/hashicorp/terraform:light
+    steps:
+      - checkout
+      - run:
+          name: terraform init
+          command: |
+            cd terraform
+            rm -rf .terraform
+            terraform init -input=false
```
Comment (wbreza) [inline .circleci/config.yml:113]: I thought you had remote state hooked up?  If so you will need to reference backend config here.

File: .circleci/config.yml [hunk 8]
```diff
@@ -80,3 +98,50 @@ jobs:
...
+              -out out.tfplan
+      - run:
+          name: terraform validate
+          command: terraform validate       
+      - persist_to_workspace:
+          root: .
+          paths:
+            - .
+
+  terraform-apply:
```
Comment (wbreza) [inline .circleci/config.yml:133]: Is the apply split out to showcase manual approval checks here?

**Review by wbreza — APPROVED** [review]
Looks good overall.  Added a few questions and comments for some items that aren't 100% clear.

File: .circleci/config.yml (hunk 6 above)
Comment (tbarlow12) [reply .circleci/config.yml:29]: Contexts are similar to variable groups in azdo. Collections of environment variables that can be shared across projects. We can use them to map variables to the target environment. (e.g. `tf-deploy-prod` could have deployment variables required to deploy to production)

File: .circleci/config.yml (hunk 8 above)
Comment (tbarlow12) [reply .circleci/config.yml:133]: Exactly

File: .circleci/config.yml [hunk 9]
```diff
@@ -80,3 +98,50 @@ jobs:
...
+            rm -rf .terraform
+            terraform init -input=false
+      - run:
+          name: terraform plan
+          command: |
+            cd terraform
+            terraform plan \
+              -var "location=${REGION}" \
+              -var "tags={environment="\"${ENV}"\"}" \
+              -var "resource_group_name=${RESOURCE_GROUP_NAME}" \
```
Comment (tessmichi) [inline .circleci/config.yml:121]: Where does the script read these variables from? Would love to be able to check a RG that was made by a run of the circle ci

Comment (wbreza) [reply .circleci/config.yml:125]: There are configured in the Circle CI `context` objects which is similar to variable groups in AzDO.

**Review by tessmichi — APPROVED** [review]
LGTM!
sidenote, if i were to change anything, it would be to add a branch other than main that also triggers a ci. but that can just go in the backlog

**Pull request merged**